File: TP1/Parte 2/Java/matrix.java
import java.util.Random;

public class MatricesAleatorias
{
    private static int rows = 5;
    private static int columns = 5;
    private static int max = 32;
    private static int min = -32;

    public static void main(String[] args) throws InterruptedException
    {

        int[][] matrixA = generateRadomMatrix(rows, columns, min, max);
        int[][] matrixB = generateRadomMatrix(rows, columns, min, max);
        int[][] matrixC = addArraysSecuencial(matrixA,matrixB);
        int[][] matrixD = addArraysWithThreads(matrixA,matrixB);

        System.out.println("matrix A:");
        printMatrix(matrixA);

        System.out.println("\nmatrix B:");
        printMatrix(matrixB);

        boolean matrixAreEquals = arraysAreEquals(matrixC,matrixD);
        if(matrixAreEquals)
        {
            System.out.println("\nmatrix CC:");
            printMatrix(matrixD);
        }
        else
        {
            System.out.println("\nLas matrices no son iguales.");
        }

    }


    public static int[][] generateRadomMatrix(int rows, int columns, int min, int max)
    {
        int[][] matrix = new int[rows][columns];
        Random rand = new Random();

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < columns; j++)
            {
                matrix[i][j] = rand.nextInt((max - min + 1)) + min;
            }
        }

        return matrix;
    }

    public static void printMatrix(int[][] matrix)
    {
        int padding = 5;
        for (int i = 0; i < matrix.length; i++)
        {
            for (int j = 0; j < matrix[i].length; j++)
            {
                String number = String.format("%" + padding + "d", matrix[i][j]);
                System.out.print(number + " ");
            }
            System.out.println();
        }
    }

    public static int[][] addArraysSecuencial(int[][] matrixA, int[][] matrixB)
    {
        int rows = matrixA.length;
        int columns = matrixA[0].length;
        int[][] result = new int[rows][columns];

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < columns; j++)
            {
                result[i][j] = matrixA[i][j] + matrixB[i][j];
            }
        }

        return result;
    }

    public static int[][] addArraysWithThreads (int[][]matrixA, int[][] matrixB) throws InterruptedException
    {
        int[][] arrayAdd = new int[rows][columns];

        Thread[] threads = new Thread[rows];

        for (int i = 0; i < rows; i++)
        {
            final int row = i;
            threads[i] = new Thread(() ->
            {
                for (int j = 0; j < columns; j++)
                {
                    arrayAdd[row][j] = matrixA[row][j] + matrixB[row][j];
                }
            }
            );
            threads[i].start();
        }

        for (Thread thread : threads)
        {
            try
            {
                thread.join();
            }
            catch (InterruptedException e)
            {
                e.printStackTrace();
            }
        }
        return arrayAdd;
    }

     public static boolean arraysAreEquals(int[][] matrixA, int[][] matrixB)
     {
        if (matrixA.length != matrixB.length || matrixA[0].length != matrixB[0].length)
        {
            return false;
        }
        for (int i = 0; i < matrixA.length; i++)
        {
            for (int j = 0; j < matrixA[0].length; j++)
            {
                if (matrixA[i][j] != matrixB[i][j])
                {
                    return false;
                }
            }
        }
        return true;
    }
}

File: TP1/Parte 2/C++/matrix.cpp
#include <thread>
#include <iostream>
#include <cstdlib>
#include <ctime>

#define MAX 32
#define MIN -32
#define ROWS 5
#define COLUMNS 5
#define THREADS 5

int GenerateRandomNumber(int min, int max);
int** GenerateMatrix(int rows, int columns);
void PrintMatrix(int **matrix, int rows, int columns, std::string name);
void FillMatrix(int** matrix, int rows, int columns);
void SumSecuencial(int** A, int** B, int** result, int rows, int columns);
void SumConcurrent(int** A, int** B, int** result, int rows, int columns, int threads);
void CompareMatrix(int **A, int **b, int rows, int columns);
void SumRow(int* A, int* B, int* result, int columns);
void FreeMatrix(int **matrix, int rows);

int main(int argc, char *argv[])
{
    srand(time(NULL));

    int **A = GenerateMatrix(ROWS, COLUMNS);
    int **B = GenerateMatrix(ROWS, COLUMNS);
    int **CS = GenerateMatrix(ROWS, COLUMNS);
    int **CC = GenerateMatrix(ROWS, COLUMNS);

    FillMatrix(A, ROWS, COLUMNS);
    FillMatrix(B, ROWS, COLUMNS);
    SumSecuencial(A, B, CS, ROWS, COLUMNS);
    SumConcurrent(A, B, CC, ROWS, COLUMNS, THREADS);

    PrintMatrix(A, ROWS, COLUMNS, "A");
    PrintMatrix(B, ROWS, COLUMNS, "B");
    PrintMatrix(CS, ROWS, COLUMNS, "CS");
    PrintMatrix(CC, ROWS, COLUMNS, "CC");

    CompareMatrix(CC, CS, ROWS, COLUMNS);

    FreeMatrix(A, ROWS);
    FreeMatrix(B, ROWS);
    FreeMatrix(CC, ROWS);
    FreeMatrix(CS, ROWS);
}


int** GenerateMatrix(int rows, int columns)
{
   int **result = (int **) calloc(rows, sizeof (int *) * rows);
   for(int i = 0; i < rows; i++)
   {
     result[i] = (int *) calloc(columns, sizeof (int) * columns);
   }
   return result;
}


void FillMatrix(int** matrix, int rows, int columns)
{
  for(int i = 0; i < rows; i++)
     for(int j= 0; j < columns; j++)
       matrix[i][j] = GenerateRandomNumber(MIN, MAX);
}


int GenerateRandomNumber(int min, int max)
{
   int mod = (((max) + 1)-(min));
   int r = rand() % mod + min;
   return r;
}


void PrintMatrix(int **matrix, int rows, int columns, std::string name)
{
     std::cout<<"\n\nMatrix "<<name<<":\n\n"<<std::endl;
     for(int i = 0; i < rows; i++)
     {
       for(int j= 0; j < columns; j++)
         printf(" %d ", matrix[i][j]);
       printf("\n");
     }
}


void SumSecuencial(int** A, int** B, int** result, int rows, int columns)
{
  for(int i = 0; i < rows; i++)
    for(int j = 0; j < columns; j++)
      result[i][j] = A[i][j] + B[i][j];
}


void SumConcurrent(int** A, int** B, int** result, int rows, int columns, int threads)
{
  std::thread v_th[threads];
  for(int i = 0; i < threads; i++)
  {
    v_th[i] = std::thread(SumRow, A[i], B[i], result[i], columns);
  }
  for(int i = 0; i < threads; i++)
  {
    v_th[i].join();
  }
}


void SumRow(int* A, int* B, int* result, int columns)
{
  for(int i = 0; i < columns; i++)
    result[i] = A[i] + B[i];
}


void FreeMatrix(int **matrix, int rows)
{
  for(int i = 0; i < rows; i++)
  {
    free(matrix[i]);
  }
  free(matrix);
}


void CompareMatrix(int **A, int **B, int rows, int columns)
{
  int iguales = 1;
  for(int i = 0; i < rows; i++)
    for(int j = 0; j < columns; j++)
      if(A[i][j] != B[i][j])
      {
         iguales = 0;
         i = rows;
         j = columns;
      }
  if(iguales)
  {
    printf("\n\nLas matrices CC y CS son IGUALES");
  }
  else
  {
    printf("\n\nLas matrices CC y CS son DISTINTAS");
  }
}
